### PR TITLE
Task views

### PR DESCRIPTION
**Pull request by cyril-steimer**

### CODE REVIEW

File: InteractionBase/src/commands/Command.cpp [hunk 1]
```diff
@@ -25,13 +25,14 @@
 ***********************************************************************************************************************/
 
 #include "commands/Command.h"
+#include "precompiled.h"
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/Command.cpp]: You never need to include the precompiled header. It's included automatically. Please remove this.

File: InteractionBase/src/commands/Command.cpp [hunk 2]
```diff
@@ -25,13 +25,14 @@
...
 
 namespace Interaction {
 
-Command::~Command()
+Command::Command(QString name, bool canBeUsedInMenu)
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/Command.cpp:31]: Why did you remove the destructor of Command? We need it here even if it's empty since it's important to put the destructors of virtual base classes in a cpp file, in order to indicate where the virtual table will be. Please bring it back.

File: InteractionBase/src/commands/Command.h [hunk 3]
```diff
@@ -39,7 +39,7 @@ namespace Interaction {
 class INTERACTIONBASE_API Command
 {
 	public:
-		virtual ~Command();
+		Command(QString name = "", bool canBeUsedInMenu = false);
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/Command.h:42]: Destructor must remain. and must still be virtual.

File: InteractionBase/src/commands/NewViewCommand.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "NewViewCommand.h"
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/NewViewCommand.cpp]: The convention for command class names is CXXXX. Please call this class and the corresponding files CNewView

File: InteractionBase/src/commands/NewViewCommand.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#include "VisualizationBase/src/items/ViewItem.h"
+
+namespace Interaction {
+
+NewViewCommand::NewViewCommand()
+	:MenuCommand("newView", QStringList({"name", ""}))
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/NewViewCommand.cpp]: I don't think you need the QStringList here. Just {"name", ""} should be sufficient.

File: InteractionBase/src/commands/NewViewCommand.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+bool NewViewCommand::canUseTarget(Visualization::Item *, Visualization::Item *,
+		const std::unique_ptr<Visualization::Cursor>&)
+{
+	return true;
+}
+
+CommandResult* NewViewCommand::executeWithArguments(Visualization::Item *, Visualization::Item *target,
+		const QStringList& arguments, const std::unique_ptr<Visualization::Cursor>&)
+{
+	QString name = arguments.at(0);
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/NewViewCommand.cpp]: Are we guaranteed to have 2 arguments here?

File: InteractionBase/src/commands/NewViewCommand.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	bool open = arguments.at(1).compare("open", Qt::CaseInsensitive) == 0;
+	auto view = new Visualization::ViewItem(nullptr, name);
+	target->scene()->addTopLevelItem(view, false);
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/NewViewCommand.cpp]: We should probably at some point have a separate method in scene that adds a top-level view.

File: InteractionBase/src/commands/RemoveNodeCommand.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+#include "VisualizationBase/src/items/ViewItem.h"
+
+namespace Interaction {
+
+RemoveNodeCommand::RemoveNodeCommand()
+	:MenuCommand("removeNode", QStringList())
+{
+}
+
+bool RemoveNodeCommand::canUseTarget(Visualization::Item *, Visualization::Item *target,
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/RemoveNodeCommand.cpp]: We should discuss the canUseTarget in our next meeting.

File: InteractionBase/src/commands/RemoveNodeCommand.cpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	return target->hasNode() && target->scene()->
+			currentViewItem()->allNodes().contains(target->node());
+}
+
+CommandResult* RemoveNodeCommand::executeWithArguments(Visualization::Item *, Visualization::Item *target,
+		const QStringList&, const std::unique_ptr<Visualization::Cursor>&)
+{
+	target->scene()->currentViewItem()->removeNode(target->node());
+	return new CommandResult();
+
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/RemoveNodeCommand.cpp]: remove this empty line

File: InteractionBase/src/commands/RemoveNodeCommand.cpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+QString RemoveNodeCommand::description(Visualization::Item *, Visualization::Item *,
+		const QStringList &, const std::unique_ptr<Visualization::Cursor> &)
+{
+	return "Remove the current top-level node from the current view";
+
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/RemoveNodeCommand.cpp]: remove empty line

File: InteractionBase/src/commands/SwitchViewCommand.cpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ **
+ **********************************************************************************************************************/
+
+#include "SwitchViewCommand.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+
+namespace Interaction {
+
+SwitchViewCommand::SwitchViewCommand()
+	:MenuCommand("switch", QStringList({"name"}))
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/SwitchViewCommand.cpp]: I don't think we need the QStringList.

File: InteractionBase/src/commands/SwitchViewCommand.cpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+bool SwitchViewCommand::canUseTarget(Visualization::Item *, Visualization::Item *,
+		const std::unique_ptr<Visualization::Cursor>&)
+{
+	return true;
+}
+
+CommandResult* SwitchViewCommand::executeWithArguments(Visualization::Item *, Visualization::Item *target,
+		const QStringList& arguments, const std::unique_ptr<Visualization::Cursor>&)
+{
+	QString name = arguments.at(0);
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/SwitchViewCommand.cpp]: are we guaranteed to have an argument?

File: InteractionBase/src/commands/SwitchViewCommand.cpp [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	auto view = target->scene()->getViewItem(name);
+	if (view)
+	{
+		target->scene()->switchToView(view);
```
Comment (dimitar-asenov) [inline InteractionBase/src/commands/SwitchViewCommand.cpp]: I am in favor of having an overload of Scene::switchToView, which directly takes a QString viewName and tries to switch if such a view exists. This will make the action easily accessible also from other code.

File: InteractionBase/src/handlers/GenericHandler.cpp [hunk 14]
```diff
@@ -115,6 +120,9 @@ GenericHandlerManagerListener& GenericHandler::managerListener()
 
 GenericHandler::GenericHandler()
 {
+	supportedCommands.append(new NewViewCommand());
```
Comment (dimitar-asenov) [inline InteractionBase/src/handlers/GenericHandler.cpp]: Actually the correct place to add these commands would be to the SceneItemHandler

File: InteractionBase/src/handlers/GenericHandler.cpp [hunk 15]
```diff
@@ -153,12 +161,55 @@ void GenericHandler::showCommandPrompt(Visualization::Item* commandReceiver, QSt
 {
 	if (commandPrompt_ && commandPrompt_->commandReceiver() == commandReceiver)
 	{
-		commandPrompt_->showPrompt(initialCommandText);
+        commandPrompt_->showPrompt(initialCommandText);
```
Comment (dimitar-asenov) [inline InteractionBase/src/handlers/GenericHandler.cpp]: Please fix the whitespace here and everywhere else. Use only TABs for indentation.

Comment (dimitar-asenov) [inline InteractionBase/src/handlers/GenericHandler.cpp]: Use a regex search string to find all issues: "^ +"

File: InteractionBase/src/handlers/GenericHandler.h [hunk 16]
```diff
@@ -85,7 +85,9 @@ class INTERACTIONBASE_API GenericHandler : public Visualization::InteractionHand
 		void removeCommandPrompt();
 		void showCommandPrompt(Visualization::Item* commandRecevier, QString initialCommandText = QString());
 
-		void toggleComment(Visualization::Item* itemWithComment, Model::Node* aNode, bool hideOnly);
+		void showCommandMenu(Visualization::Item* commandReceiver);
+
+        void toggleComment(Visualization::Item* itemWithComment, Model::Node* aNode, bool hideOnly);
```
Comment (dimitar-asenov) [inline InteractionBase/src/handlers/GenericHandler.h]: white space

File: InteractionBase/test/SimpleTest.cpp [hunk 17]
```diff
@@ -34,7 +34,7 @@
 #include "VisualizationBase/src/VisualizationManager.h"
 #include "VisualizationBase/src/items/VList.h"
 #include "VisualizationBase/src/items/VComposite.h"
-#include "VisualizationBase/src/items/RootItem.h"
+#include "VisualizationBase/src/Scene.h"
```
Comment (dimitar-asenov) [inline InteractionBase/test/SimpleTest.cpp]: This is already included four lines above. Please remove it here.

File: InteractionBase/test/SimpleTest.cpp [hunk 18]
```diff
@@ -73,12 +73,11 @@ TEST(InteractionBasePlugin, TextSelect)
...
 	manager->endModification();
 
-	auto top = new RootItem(list);
 	auto scene = VisualizationManager::instance().mainScene();
-	scene->addTopLevelItem( top );
+	scene->addTopLevelNode( list );
 	QApplication::processEvents();
 
-	VList* list2 = dynamic_cast<VList*> (top->wrappedItem());
+	VList* list2 = dynamic_cast<VList*> (list);
```
Comment (dimitar-asenov) [inline InteractionBase/test/SimpleTest.cpp]: This is never going to work. list is a Node and here we're casting to something of type item.

You need to provide a way to get to the item. You can for example get the current viewItem from the scene and then use the findVisualizationOf method to find the corresponding item to the list Node.

File: OOInteraction/src/OOInteractionPlugin.cpp [hunk 19]
```diff
@@ -162,6 +163,8 @@ bool OOInteractionPlugin::initialize(Core::EnvisionManager&)
 	OOInteraction::DoxygenWholeTreeVisitor::init();
 	OOInteraction::DoxygenCommentsOnlyVisitor::init();
 
+	Interaction::GenericHandler::instance()->addCommand(new AddNodeCommand());
```
Comment (dimitar-asenov) [inline OOInteraction/src/OOInteractionPlugin.cpp]: I doubt we want to include this for all node types. Some things don't make sense, e.g. a single literal.

For now why don't we register this for all declarations only?

File: OOInteraction/src/OOInteractionPlugin.cpp [hunk 20]
```diff
@@ -49,6 +49,7 @@
 #include "commands/CCreateMethod.h"
 #include "commands/CSceneHandlerItemTest.h"
 #include "commands/CDoxygen.h"
+#include "commands/AddNodeCommand.h"
```
Comment (dimitar-asenov) [inline OOInteraction/src/OOInteractionPlugin.cpp]: Please change the AddNodeCommand class to follow the right conventions CXXX.

Also AddNode is too unspecific. Use CAddNodeToView

File: OOInteraction/src/commands/AddNodeCommand.cpp [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+ **********************************************************************************************************************/
+
+#include "AddNodeCommand.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+#include "OOModel/src/declarations/Project.h"
+
+namespace OOInteraction {
+
+AddNodeCommand::AddNodeCommand()
+	:MenuCommand("addNode", QStringList({"current", "0", "0"}))
```
Comment (dimitar-asenov) [inline OOInteraction/src/commands/AddNodeCommand.cpp]: remove QStringList

File: OOInteraction/src/commands/AddNodeCommand.cpp [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+	{
+		view->insertNode(method, column, row);
+		return new Interaction::CommandResult();
+	}
+	else if (!view)
+		return new Interaction::CommandResult(new Interaction::CommandError(
+											"The view with name " + name + " does not exist"));
+	else if (!colOk)
+		return new Interaction::CommandResult(new Interaction::CommandError(
+												"Could not parse " + arguments.at(1) + " to integer"));
```
Comment (dimitar-asenov) [inline OOInteraction/src/commands/AddNodeCommand.cpp]: A better and shorter message here  and below is"
arguments.at(1) + " is not an integer"

File: OOInteraction/src/commands/AddNodeCommand.cpp [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+{
+	return target->hasNode() && dynamic_cast<OOModel::Method*>(target->node());
+}
+
+Interaction::CommandResult* AddNodeCommand::executeWithArguments(Visualization::Item *, Visualization::Item *target,
+		const QStringList& arguments, const std::unique_ptr<Visualization::Cursor>&)
+{
+	auto method = dynamic_cast<OOModel::Method*>(target->node());
+	auto name = arguments.at(0);
+	auto colOk = true;
```
Comment (dimitar-asenov) [inline OOInteraction/src/commands/AddNodeCommand.cpp]: I believe there should be a quick programatic way of adding a node in a specific row and column, but I doubt that this is something the user would want to specify manually on the command line. Do you think it's necessary?

File: OOInteraction/src/commands/CSceneHandlerItemTest.cpp [hunk 24]
```diff
@@ -26,7 +26,7 @@
 
 #include "CSceneHandlerItemTest.h"
 
-#include "VisualizationBase/src/items/RootItem.h"
+#include "VisualizationBase/src/Scene.h"
```
Comment (dimitar-asenov) [inline OOInteraction/src/commands/CSceneHandlerItemTest.cpp]: You don't need this as it is already automatically included by some of the other includes here.

File: OOInteraction/src/handlers/HMethod.cpp [hunk 25]
```diff
@@ -37,10 +37,14 @@
...
 #include "VisualizationBase/src/cursor/LayoutCursor.h"
 
+#include "commands/AddNodeCommand.h"
+
 namespace OOInteraction {
 
 HMethod::HMethod()
-{}
+{
+	addCommand(new AddNodeCommand());
```
Comment (dimitar-asenov) [inline OOInteraction/src/handlers/HMethod.cpp]: Didn't you already add this to GenericHandler? Why do we add it again here?

File: VisualizationBase/src/Scene.cpp [hunk 26]
```diff
@@ -109,11 +112,24 @@ ModelRenderer* Scene::defaultRenderer()
 	return &defaultRenderer_;
 }
 
-void Scene::addTopLevelItem(Item* item)
+void Scene::addTopLevelNode(Node *node, int column, int row)
+{
+	currentViewItem()->insertNode(node, column, row);
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp:117]: We should make sure that there is at least one view. If not we should add one.

File: VisualizationBase/src/Scene.cpp [hunk 27]
```diff
@@ -109,11 +112,24 @@ ModelRenderer* Scene::defaultRenderer()
...
+}
+
+void Scene::addTopLevelItem(Item* item, bool show)
 {
 	Q_ASSERT(!inAnUpdate_);
+	//If we have a ViewItem, we must add it to all ViewItems
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp:123]: In the new code, does this ever happen? We should not allow this.

Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp:123]: I.E. Views should not be added via addTopLevelItem to the viewItems_ but rather adding to the viewItems should call addTopLevelItem which should work as before.

File: VisualizationBase/src/Scene.cpp [hunk 28]
```diff
@@ -126,6 +142,38 @@ void Scene::removeTopLevelItem(Item* item)
 	scheduleUpdate();
 }
 
+
+ViewItem* Scene::currentViewItem()
+{
+	if (!currentViewItem_)
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp:148]: hmm, we should discuss whether this should happen here or somewhere else.

File: VisualizationBase/src/Scene.cpp [hunk 29]
```diff
@@ -126,6 +142,38 @@ void Scene::removeTopLevelItem(Item* item)
...
+		if (viewItems_.size() == 0)
+			addTopLevelItem(new ViewItem(nullptr, "ProjectView"));
+		else
+			currentViewItem_ = viewItems_.first();
+	}
+	Q_ASSERT(currentViewItem_);
+	return currentViewItem_;
+}
+
+ViewItem* Scene::getViewItem(QString name)
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp]: By convention we do not prefix getters with get. Rename this to viewItem

File: VisualizationBase/src/Scene.cpp [hunk 30]
```diff
@@ -126,6 +142,38 @@ void Scene::removeTopLevelItem(Item* item)
...
+{
+	for (auto item : viewItems_)
+		if (item->name() == name)
+			return item;
+	return nullptr;
+}
+
+void Scene::switchToView(ViewItem *view)
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.cpp:167]: There should be an overload that takes a string name and returns true or false if it managed to switch.

File: VisualizationBase/src/Scene.h [hunk 31]
```diff
@@ -65,12 +66,19 @@ class VISUALIZATIONBASE_API Scene : public QGraphicsScene
...
-		void addTopLevelItem(Item* item);
+		void addTopLevelNode(Node* node, int column = 0, int row = 0);
+		void addTopLevelItem(Item* item, bool show = true);
 		void removeTopLevelItem(Item* item);
 		void scheduleUpdate();
 		void updateNow();
 		void listenToTreeManager(Model::TreeManager* manager);
 
+		ViewItem* getViewItem(QString name);
+
```
Comment (dimitar-asenov) [inline VisualizationBase/src/Scene.h]: don't leave empty spaces between the methods that handle the viewItems

File: VisualizationBase/src/items/ViewItem.cpp [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+	for (int i = 0; i < nodes_.size(); i++)
+	{
+		auto index = nodes_.at(i).indexOf(node);
+		if (index != -1)
+			nodes_[i].remove(index);
+	}
+	scene()->scheduleUpdate();
+}
+
+const QList<Model::Node*> ViewItem::allNodes() const
```
Comment (dimitar-asenov) [inline VisualizationBase/src/items/ViewItem.cpp:73]: Where do you need this method?

File: InteractionBase/src/commands/NewViewCommand.cpp (hunk 6 above)
Comment (cyril-steimer) [inline InteractionBase/src/commands/NewViewCommand.cpp]: Yes, the MenuCommand will automatically fill up non-existing arguments with the default.

File: InteractionBase/src/commands/SwitchViewCommand.cpp (hunk 12 above)
Comment (cyril-steimer) [inline InteractionBase/src/commands/SwitchViewCommand.cpp]: Yes, empty arguments are automatically filled in.

File: OOInteraction/src/commands/AddNodeCommand.cpp (hunk 23 above)
Comment (cyril-steimer) [inline OOInteraction/src/commands/AddNodeCommand.cpp]: As empty arguments are filled in by default, not specifying them will simply use the default arguments 0. As such, the user does not need to, but can specifiy it in the command.

File: VisualizationBase/src/items/ViewItem.cpp (hunk 32 above)
Comment (cyril-steimer) [inline VisualizationBase/src/items/ViewItem.cpp:73]: I use it in the RemoveNodeCommand.
